### ISSUE TITLE
Server response time increased after moving to SocketCluster

### ISSUE THREAD

**Issue by snegostup** [SocketCluster/socketcluster#153]
I have a website, it uses Express as a web server. Before it served a page in ~0.03-0.05 sec. After I moved to SocketCluster (1 broker 4 workers) it serves the page in ~0.2-0.3 sec.
I understand that SocketCluster brings some overhead compared to raw 1 process Express. My question is is it normal for load time to increase THAT much?

**Comment by jondubois** [comment]
I can think of only three things which SC does which could add overhead to the HTTP server:
- The Node.js cluster module which distributes requests between processes
- The WS module (https://github.com/websockets/ws) which interfaces with the HTTP server
- Then there is this code which adds some custom SC properties to the req object

Maybe the Node.js cluster module may be adding extra latency? I wouldn't be too worried about this just yet; a 6x increase in latency doesn't necessarily mean a 6x increase in CPU load (especially since these are still all quite small numbers) - If it does increase CPU load, then that's definitely something that needs to be investigated and fixed.

I haven't noticed any performance issues with SC based on the stress testing I did, but then those tests were mostly focused on WebSocket connections/data and not HTTP.

**Comment by jondubois** [comment]
The other thing to consider is that the default SC boilerplate uses the ```serve-static``` module  with express - I'm not sure what sort of IO caching it uses on the server-side, if you use a different module, you may get different results.

**Comment by jondubois** [comment]
I did a test serving a simple .js file over HTTP and I'm getting the same latency for Express with SC and for plain Express without SC. I did notice that there is a huge difference between the first time the file is loaded vs subsequent times - It took 6x to 10x as long on subsequent calls (sometimes more). This is related to caching.

Caching happens in both the browser and on the server. Even if you disable browser caching, it will still be faster the second time you fetch a particular file because of server-side caching. If you restart the server, it will clear the server-side cache.

Note that if you use SC with 4 workers, each worker needs to get its own copy of the cached file so that means that after starting up the server, the cache could miss up to 4 times if the request ends up on a different worker each time. On a plain single-process express server, the cache will miss once at most.
